In [1]:
import time
import json
import pandas as pd
from openai import AzureOpenAI
from threading import Semaphore
from concurrent.futures import ThreadPoolExecutor, as_completed
from typing import List, Dict, Any
from tenacity import (
    retry,
    stop_after_attempt,
    wait_fixed,
    wait_random,
)

# Data Format Example

The input data, `message_sequences`, should be structured as a list of message dictionaries, where each message has a `role` (either "system" or "user") and `content`. Here’s a very simple example:

In [2]:
message_sequences = []
for i in range(1, 101):
    message_sequences.append([
    {"role": "system", "content": "You are a helpful AI assistant.  Always answer in the following json format: {'content': '2'}."},
     {"role": "user", "content": f"What is {i} + {i}?"}
    ])
message_sequences[:3]

[[{'role': 'system',
   'content': "You are a helpful AI assistant.  Always answer in the following json format: {'content': '2'}."},
  {'role': 'user', 'content': 'What is 1 + 1?'}],
 [{'role': 'system',
   'content': "You are a helpful AI assistant.  Always answer in the following json format: {'content': '2'}."},
  {'role': 'user', 'content': 'What is 2 + 2?'}],
 [{'role': 'system',
   'content': "You are a helpful AI assistant.  Always answer in the following json format: {'content': '2'}."},
  {'role': 'user', 'content': 'What is 3 + 3?'}]]

In [3]:
class APIRequestor:
    def __init__(self, model_name, temperature=1.0, max_tokens=15, rate_limit=20):
        self.client = AzureOpenAI(
                # https://learn.microsoft.com/en-us/azure/ai-services/openai/reference#rest-api-versioning
                api_version="2023-07-01-preview",
                # https://learn.microsoft.com/en-us/azure/cognitive-services/openai/how-to/create-resource?pivots=web-portal#create-a-resource
               azure_endpoint= "https://chm-openai.openai.azure.com/",
            )
        self.model_name = model_name
        self.temperature = temperature
        self.max_tokens = max_tokens
        self.rate_limit = rate_limit  # max requests per minute, adjust according to your limit
        self.semaphore = Semaphore(self.rate_limit)
    def handle_last_retry_error(retry_state):
        print(f"All retry attempts failed for: {retry_state.args[0]}\nReturning None for this request.")
        return '{"content": "None"}'  # Custom behavior after all retries fail

    @retry(wait=wait_fixed(2) + wait_random(0, 2),
            stop=stop_after_attempt(2),
            before_sleep= lambda retry_state: print("Retrying..."),
            retry_error_callback=handle_last_retry_error)
    def get_response(self, system_user_message: List):
            with self.semaphore:
                try:
                    response = self.client.chat.completions.create(
                                model=self.model_name,
                                messages= system_user_message,
                                temperature= self.temperature,
                                max_tokens=self.max_tokens,
                                response_format= { "type": "json_object" },
                            )
                    json_response = response.choices[0].message.content
                    json.loads(json_response)  # Attempt to parse the JSON, will raise an error if not valid JSON
                    time.sleep(60 / self.rate_limit)  # Pause to respect the rate limit to evenly distribute requests over time
                    return json_response
                except json.JSONDecodeError:
                    print(f"Invalid JSON response for message {system_user_message}")
                    raise  # This re-raises the last exception, triggering a retry
                except Exception as e:
                    print(f"Error while processing: {str(e)}")
                    raise
         
    def get_responses_parallel(self, messages_list):
        results = []
        started = time.time()
        with ThreadPoolExecutor(max_workers=self.rate_limit) as executor:
            future_to_message = {executor.submit(self.get_response, message): message for message in messages_list}
            for future in as_completed(future_to_message):
                message = future_to_message[future]
                try:
                    result = future.result()
                    # Extract the user's query from the message (response format is a JSON string like '{"content": "4"}')
                    user_query = next((msg['content'] for msg in message if msg['role'] == 'user'), "Unknown query")
                    result_content = json.loads(result)['content'] # if result is 
                    results.append({"input": user_query, "content": result_content})
                except Exception as e:
                      results.append({"input": "Error", "content": f"Error processing message: {str(e)}"})
        print(f"Total time taken: {time.time() - started} seconds")
        return results               

Ensure you stay within the rate limit for the selected model

In Our Azure AI settings, gpt-35-turbo has a Rate limit (Requests per minute) of 720

In [4]:
gpt35_turbo_api = APIRequestor(model_name = "gpt-35-turbo", temperature = 1.0, max_tokens = 20, rate_limit = 100) 

In [5]:
results_parallel = gpt35_turbo_api.get_responses_parallel(message_sequences)

Total time taken: 2.5722763538360596 seconds


In [6]:
results_parallel[:5]

[{'input': 'What is 53 + 53?', 'content': '106'},
 {'input': 'What is 100 + 100?', 'content': '200'},
 {'input': 'What is 42 + 42?', 'content': '84'},
 {'input': 'What is 56 + 56?', 'content': '112'},
 {'input': 'What is 19 + 19?', 'content': '38'}]

As u can notice, the order of the responses is not the same as the order of the queries, because the requests are processed in parallel

In [7]:
df_parallel = pd.DataFrame(results_parallel)
df_parallel.to_csv("results_parallel.csv", index=False)
df_parallel

input content
0     What is 53 + 53?     106
1   What is 100 + 100?     200
2     What is 42 + 42?      84
3     What is 56 + 56?     112
4     What is 19 + 19?      38
..                 ...     ...
95    What is 38 + 38?      76
96    What is 45 + 45?      90
97    What is 65 + 65?     130
98    What is 48 + 48?      96
99    What is 49 + 49?      98

[100 rows x 2 columns]

### Now a one-per-one request method to compare it with the parallel method

In [8]:
class APIRequestor_simpler:
    def __init__(self, model_name, temperature=1.0, max_tokens=15):
        self.client = AzureOpenAI(
                # https://learn.microsoft.com/en-us/azure/ai-services/openai/reference#rest-api-versioning
                api_version="2023-07-01-preview",
                # https://learn.microsoft.com/en-us/azure/cognitive-services/openai/how-to/create-resource?pivots=web-portal#create-a-resource
               azure_endpoint= "https://chm-openai.openai.azure.com/",
            )
        self.model_name = model_name
        self.temperature = temperature
        self.max_tokens = max_tokens

    def get_response(self, system_user_message: List):
                try:
                    response = self.client.chat.completions.create(
                                model=self.model_name,
                                messages= system_user_message,
                                temperature= self.temperature,
                                max_tokens=self.max_tokens,
                                response_format= { "type": "json_object" },
                            )
                    json_response = response.choices[0].message.content
                    json.loads(json_response)  # Attempt to parse the JSON, will raise an error if not valid JSON
                    return json_response
                except json.JSONDecodeError:
                    print(f"Invalid JSON response for message {system_user_message}\nReturning None for this request.")
                    return '{"content": "None"}'  # This re-raises the last exception, triggering a retry
                except Exception as e:
                    print(f"Error while processing: {str(e)}")
                    raise
                 

In [9]:
gpt35_turbo_api2 = APIRequestor_simpler(model_name = "gpt-35-turbo", temperature = 1.0, max_tokens = 20)

In [10]:
started = time.time()
results_2 = []
for i in message_sequences:
    results_2.append(gpt35_turbo_api2.get_response(i))
print(f"Time taken: {time.time() - started}")


Time taken: 18.58565092086792


#### As you can notice, processing these requests one by one took about 18.6 seconds. However, using the parallel processing method, this time was significantly reduced to approximately 2.6 seconds, making it 7 times faster.

